src/AddressDerivation.hs

```diff
@@ -0,0 +1,215 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RoleAnnotations #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module AddressDerivation
+    (
+    -- * Overview
+    -- $overview
+
+    -- * Abstractions
+      GenMasterKey (..)
+    , HardDerivation (..)
+    , SoftDerivation (..)
+
+    -- * Helper types
+    , Index (..)
+    , Depth (..)
+    , DerivationType (..)
+
+    ) where
+
+import Prelude
+
+import Cardano.Crypto.Wallet
+    ( XPrv, XPub )
+import Cardano.Mnemonic
+    ( SomeMnemonic )
+import Control.DeepSeq
+    ( NFData )
+import Data.ByteArray
+    ( ScrubbedBytes )
+import Data.String
+    ( fromString )
+import Data.Typeable
+    ( Typeable )
+import Data.Word
+    ( Word32 )
+import Fmt
+    ( Buildable (..) )
+import GHC.Generics
+    ( Generic )
+
+-- $overview
+--
+-- These abstractions when combined together enable address derivation
```

Comment:
That's quite a quick overview :smile:

Reply:
in this PR was focusing on the code. hope to spend more time on docs in the next PRs